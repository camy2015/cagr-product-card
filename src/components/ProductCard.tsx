import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import React, { CSSProperties, createContext } from 'react';
import {
  InitialValues,
  Product,
  ProductCardHandlers,
  ProductContextProps,
  onChangeArgs,
} from '../interfaces/interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

interface Props {
  product: Product;
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    increaseBy,
    counter,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
