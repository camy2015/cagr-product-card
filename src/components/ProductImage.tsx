import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

interface Props {
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ className, style }: Props) => {
  const { product } = useContext(ProductContext);
  const { img = '' } = product;

  return (
    <img
      src={img ? img : noImage}
      alt="Product"
      className={`${styles.productImg} ${className}`}
      style={style}
    />
  );
};
