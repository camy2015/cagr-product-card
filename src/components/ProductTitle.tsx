import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

interface Props {
  className?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ className, style }: Props) => {
  const { product } = useContext(ProductContext);
  const { id, title } = product;

  return (
    <>
      <span
        className={`${styles.productDescription} ${className}`}
        style={style}
      >
        {title}
      </span>
      <span
        className={`${styles.productDescription} ${className}`}
        style={style}
      >
        {id}
      </span>
    </>
  );
};
