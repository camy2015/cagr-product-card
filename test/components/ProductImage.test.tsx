import { describe, test, expect } from '@jest/globals';
import { ProductCard, ProductImage } from '../../src/components';
import renderer from 'react-test-renderer';
import { product2 } from '../data/products';
import React from 'react';

describe('Tests in <ProductImage/>', () => {
  test('should display the component with the image of product', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
