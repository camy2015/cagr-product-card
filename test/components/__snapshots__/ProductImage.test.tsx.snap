// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <ProductImage/> should display the component with the image of product 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
